# CRUK Bioinformatics Summer School 2018: Functional Genomics

## 23rd - 27th July 2018: Craik-Marshall Room, Downing Site, University of Cambridge


# Overview

Functional genomics looks at the dynamic aspects of how the genome functions within cells,
particularly in the form of gene expression (transcription) and gene regulation. This workshop surveys
current methods for functional genomics using high-throughput technologies. 

High-throughput technologies such as next generation sequencing (NGS) can routinely produce massive amounts of data. However, such datasets pose new challenges in the way the data have to be analyzed, annotated and interpreted which are not trivial and are daunting to the wet-lab biologist. This course covers state-of-the-art and best-practice tools for bulk RNA-seq and ChIP-seq data analysis, and will also introduce approaches for analysing data arising from single-cell RNA-seq studies.

# Audience

Enthusiastic and motivated wet-lab biologists who want to gain more of an understanding of NGS data and eventually progress to analysing their own data

# Pre-requisites

The course will include a great deal of hands-on work in R and at the command line. In order for you to make the most of the course we strongly recommend that you take an introductory course, or have sufficient experience in the following areas:

- R
 - Unix
 - Introductory statistics

More specific requirements and references can be found [here](http://www.cruk.cam.ac.uk/bioinformatics-summer-school-prerequisites)


# Instructors

- Mark Fernandes (CRUK CI)
- Shamith Samarajiwa (MRC CU)
- Dora Bihary (MRC CU)
- Ashley Sawle (CRUK CI)
- Alistair Martin (CRUK CI)
- Stephane Ballereau (CRUK CI)
- Michael Morgan (CRUK CI)
- Abigail Edwards (CRUK CI)


# Aims
During this course you will learn about:-

- To provide an understanding of how aligned sequencing reads, genome sequences and genomic regions are represented in R.
- To encourage confidence in reading sequencing reads into R, performing quality assessment and executing standard pipelines for (bulk) RNA-Seq and ChIP-Seq analysis 
- Analysis of transcription factor (TF) and epigenomic (histone mark) ChIP-seq data 
- Recent advances in single-cell sequencing

# Objectives
After the course you should be able to:-

- Know what tools are available in Bioconductor for HTS analysis and understand the basic object-types that are utilised.
- Process and quality control short read sequencing data 
- Given a set of gene identifiers, find out whereabouts in the genome they are located, and vice-versa 
- Produce a list of differentially expressed genes from an RNA-Seq experiment.
- Import a set of ChIP-Seq peaks and investigate their biological context.
- Appreciate the differences between bulk and single-cell RNA-seq analyses, and why the same methodologies might not be applicable

# Accomodation

If required, free bed and breakfast accommodation will be provided for attendees in Murray-Edwards College. Please let us know on the registration form if you need accommodation and when you plan to check-in and check-out. Please note that the course ends at 13:30 on Friday 22nd September and accommodation is not provided for Friday night.

# Timetable (approximate)


## Day zero

Informal get-together the day before the course at a local pub

## Day one 09:30 - 17:00

- Overview of Functional Genomics
- Review of the R programming language
- Data Processing for Next Generation Sequencing
- Alignment to reference genomes and QC

## Day two 09:30 - 17:00

- Introduction to (bulk) RNA-seq
- Importing and QC of RNA-seq data
- Differential Expression

## Day three 09:30 - 17:00

- Annotation and visualisation of Differential Expression
- Gene set Analysis and Gene Ontology testing
- Introduction to ChIP-seq
- Quality control methods for ChIP-seq


## Day four 09:30 - 17:00

- Downloading public ChIP-seq datasets
- Downstream analysis of ChIP-seq data
- Differential Binding analysis
- Introduction to the analysis of long distance interactions

## Day five 09:30 - 12:30

- Introduction to single-cell sequencing 
